// electron-updater/out/GitHubProvider.js
"use strict";

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.GitHubProvider = undefined;

var _bluebirdLstC;

function _load_bluebirdLstC() {
    return _bluebirdLstC = require("bluebird-lst-c");
}

var _api;

function _load_api() {
    return _api = require("./api");
}

var _GenericProvider;

function _load_GenericProvider() {
    return _GenericProvider = require("./GenericProvider");
}

var _path = _interopRequireWildcard(require("path"));

var _electronBuilderHttp;

function _load_electronBuilderHttp() {
    return _electronBuilderHttp = require("electron-builder-http");
}

var _CancellationToken;

function _load_CancellationToken() {
    return _CancellationToken = require("electron-builder-http/out/CancellationToken");
}

function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) newObj[key] = obj[key]; } } newObj.default = obj; return newObj; } }

class GitHubProvider extends (_api || _load_api()).Provider {
    constructor(options) {
        super();
        this.options = options;
    }
    getLatestVersion() {
        var _this = this;

        return (0, (_bluebirdLstC || _load_bluebirdLstC()).coroutine)(function* () {
            const basePath = _this.getBasePath();
            let version;
            const cancellationToken = new (_CancellationToken || _load_CancellationToken()).CancellationToken();
            try {
                // do not use API to avoid limit
                const releaseInfo = yield (0, (_electronBuilderHttp || _load_electronBuilderHttp()).request)({
                    hostname: "github.com",
                    path: `${basePath}/latest`,
                    headers: Object.assign({ Accept: "application/json" }, _this.requestHeaders)
                }, cancellationToken);
                version = releaseInfo.tag_name.startsWith("v") ? releaseInfo.tag_name.substring(1) : releaseInfo.tag_name;
            } catch (e) {
                throw new Error(`Unable to find latest version on github, please ensure a production release exists: ${e.stack || e.message}`);
            }
            let result;
            const channelFile = (0, (_api || _load_api()).getChannelFilename)((0, (_api || _load_api()).getDefaultChannelName)());
            const channelFileUrlPath = `${basePath}/download/v${version}/${channelFile}`;
            try {
                result = yield (0, (_electronBuilderHttp || _load_electronBuilderHttp()).request)({ hostname: "github.com", path: channelFileUrlPath, headers: _this.requestHeaders || undefined }, cancellationToken);
            } catch (e) {
                if (e instanceof (_electronBuilderHttp || _load_electronBuilderHttp()).HttpError && e.response.statusCode === 404) {
                    throw new Error(`Cannot find ${channelFile} in the latest release artifacts: ${e.stack || e.message}`);
                }
                throw e;
            }
            (0, (_GenericProvider || _load_GenericProvider()).validateUpdateInfo)(result);
            if ((0, (_api || _load_api()).getCurrentPlatform)() === "darwin") {
                result.releaseJsonUrl = `https://github.com${channelFileUrlPath}`;
            }
            return result;
        })();
    }
    getBasePath() {
        return `/${this.options.owner}/${this.options.repo}/releases`;
    }
    getUpdateFile(versionInfo) {
        var _this2 = this;

        return (0, (_bluebirdLstC || _load_bluebirdLstC()).coroutine)(function* () {
            if ((0, (_api || _load_api()).getCurrentPlatform)() === "darwin") {
                return versionInfo;
            }
            const basePath = _this2.getBasePath();
            // space is not supported on GitHub
            const name = versionInfo.githubArtifactName || _path.posix.basename(versionInfo.path).replace(/ /g, "-");
            return {
                name: name,
                url: `https://github.com${basePath}/download/v${versionInfo.version}/${name}`,
                sha2: versionInfo.sha2
            };
        })();
    }
}
exports.GitHubProvider = GitHubProvider; //# sourceMappingURL=GitHubProvider.js.map
